Logging-Tracing-Monitoring capability vision (#137)
kyma-project/community: capabilities/logging-tracing-monitoring.md

To support application observability and follow the batteries included philosophy, Kyma would provide lightweight and cloud-native solutions for logging, tracing and monitoring. Such solutions enable developers and operators to easily query all application health data across the different workloads in development and production environments.

This includes the following features:
* Setting up and maintaining optional, lightweight, and cloud-native solutions for logging, tracing and monitoring.
* As much as possible data derived out of the box from Kubernetes and the Service Mesh.
* Support for local development with Minikube.
* Pluggability of all pre-bundled tooling by having well defined interfaces in place.
* Ready-to use adapters to integrate easily with the tooling provided by the cloud providers, especially DynaTrace for monitoring.
* Easily accessible and secure API and UI/CLI to query logs, traces and metrics.
* Namespace separation support.
* Support of transactional event logs such as audit logs.
* Support for auto-scaling based on application metrics.

> * Proactive and reactive alerting for potential unhealthy cluster components

Reviewer: ```suggestion
* Proactive and reactive alerting for potential unhealthy cluster components.
```

> * Pre-integration with notification systems for alerting like VictorOps and Slack.

Reviewer: ```suggestion
* Pre-integration with notification systems for alerting, such as VictorOps and Slack.
```

---
displayName: Logging-Tracing-Monitoring
epicsLabels:
  - area/logging
  - area/tracing
  - area/monitoring
---

 ## Scope
Enable the operator and developer to easily observe the state of the Kyma cluster and distributed applications running on Kyma. The pre-bundled infrastructure exposes and collects the data through application logs and metrics as well as transaction traces. This infrastructure integrates with the cloud provider-specific tools.

> Ensuring that Kyma components create application logs, metrics, and trace data in a consistent way and on a sufficient detail level is not in scope of this capability.

Reviewer 2: I am not sure if it should be outside. At least in case of tracing consistency is essential. Cannot imagine if every component uses different standards. In my opinion, the scope should contain all the guidelines and the unifications. It would be nice if at least kyma components use the same logging levels and the same configuration mechanism for changing it.

Author: So I removed the section and keep it unspecific. Let's see at a later time who at the end will drive the contribution guidelines itself and who will check that they are followed

 ## Vision
In a cloud-native microservice architecture, a user request often flows through dozens of different microservices. Tools such as logging and monitoring provide insights on the health and state of a particular component. Based on the results, you can act proactively and reactively to keep or recover the health of a component. Tracing enriches observability by identifying transaction traces across the different components.

To support application observability and follow the batteries included philosophy, Kyma would provide lightweight and cloud-native solutions for logging, tracing and monitoring. Such solutions enable developers and operators to easily query all application health data across the different workloads in development and production environments.

> This includes:

Reviewer: ```suggestion
This includes the following features and benefits:
```

> * Set up and maintain optional, lightweight and cloud-native solutions for logging, tracing and monitoring.

Reviewer: ```suggestion
* Setting up and maintaining optional, lightweight, and cloud-native solutions for logging, tracing and monitoring.
```

* As much as possible data derived out of the box from Kubernetes and the Service Mesh.

> * Support for local development (Minikube).

Reviewer: ```suggestion
* Support for local development with Minikube.
```

This one may be a bit tricky, so if you don't like the expanded version we can use `(Minikube)` as it was in the previous version.

> * Pluggability of all pre-bundled tooling by having well defined interfaces in place

Reviewer: ```suggestion
* Pluggability of all pre-bundled tooling by having well defined interfaces in place.
```

* Ready-to use adapters to integrate easily with the tooling provided by the cloud providers, especially DynaTrace for monitoring.
* Easily accessible and secure API and UI/CLI to query logs.
* Namespace separation support.

> * (Logging) Support of transactional event logs such as audit logs.

Reviewer 2: Please remove (Logging) (Monitoring) prefixes.

Author: done

* (Monitoring) Support for auto-scaling based on application metrics.
* (Monitoring) Pre-integration with notification systems like VictorOps and Slack.

> This enables the following:

Reviewer: ```suggestion
The features enable the following:
```

> The features enable the following:

Reviewer 2: I would remove that section and include mentioned here benefits as features. Example:
* Built-in tracing of distributed applications based on microservices including eventing and external systems connectivity. (first 2 points)
*  Pre-integration with notification systems like VictorOps and Slack for alerts about unhealthy cluster components (last 2 points)

Author: I removed the section and applied the points as features

* Tracing the distributed transaction as it propagates across components and microservices.
* Tracing end-to-end Event flows including external applications.

> * Improved debugging in local and remote development.

Reviewer 2: Improved debugging is not clear. Do you mean something like telepresence?

Author: I meant the general improvement of debuggability by having that features in place, not specific. So I removed the sentence fully.

---
displayName: Logging-Tracing-Monitoring
epicsLabels:
  - area/logging
  - area/tracing
  - area/monitoring
---

 ## Scope
Enable operator and developer to easily observe the state of a kyma cluster and the state of distributed applications running on kyma; The data is made accessible and collectable by pre-bundled infrastructure for collecting application logs and application metrics as well as transaction traces. That infrastructure integrates with cloud provider specific tooling.

The assurance that kyma components are creating application logs and metrics as well as trace data in a consistent way and in a sufficent detail level is not in scope of the capability.

 ## Vision

> In a cloud-native microservices architecture, a request triggered by some end user action often flows through dozens or maybe more microservcies. Tooling such as logging and monitoring will provide insights about the health and state of a particular component. On base of that, pro-active and re-active actions can be taken to keep or recover the health of a component. Tracing will enrich that observability picture by adding the facet of transactions traces across the different components.

Author: Distinguish is going in the correct direction, I changed it to "identify" for now. What do you think about "isolate"?

To support observability of applications and following the *batteries-included* principle, kyma should provide leightweight and cloud-native solutions for logging, tracing and monitoring. They should enable developers and operators to easily query all application health data across the different workloads, in an development environment as well as on productive setups.

This includes:
* Setup and maintain optional, lightweight and cloud-native solutions for logging, tracing and monitoring
* As much as possible data derived out-of-the-box from Kubernetes and the Service Mesh
* Support for local development (minikube)

> * Integration of all 3 aspects into solutions of the supported cloud providers, especially DynaTrace for monitoring

Author: I meant the big topics: logging,tracing,monitoring

> * Integration of all 3 aspects into solutions of the supported cloud providers, especially DynaTrace for monitoring

Author: I changed the wording fully and spitted the point into two lines

* Easy accessible and secure API and UI/CLI to query logs
* Namespace separation support
* (Logging) Support of transactional event logs (like audit logs)
* (Monitoring) Support for auto-scaling on base of application metrics
* (Monitoring) Pre-integration with typical notification systems like VictorOps and Slack

This should enable:

> * Trace the distributed transaction as it propagetes across various components and microservices

Reviewer: ```suggestion
* Tracing the distributed transaction as it propagates across components and microservices.
```

> * Trace end-to-end event flows including outer-kyma applications

Reviewer: ```suggestion
* Tracing end-to-end Event flows including external applications.
```

Can we have `external` here, of it should mention Kyma directly?

> * Identification of culprit components for a failed distributed transaction

Reviewer: ```suggestion
* Identification of culprit components for a failed distributed transaction.
```

> * Identification of latency and performance bottlenecks along the request flow

Reviewer: ```suggestion
* Identification of latency and performance bottlenecks in the request flow.
```

> * Better debuggability for local and remote development

Reviewer: ```suggestion
* Improved debugging in local and remote development.
```

> * Pro-active notification of unhealthy states

Reviewer: ```suggestion
* Proactive notification for unhealthy states.
```

> * Reactive notification for culprit components

Reviewer: ```suggestion
* Reactive warning about culprit components.
```

I'm wondering about the word `culprit` - would `affected` be ok, or we need to indicate that the components are guilty of what is happening? if so I will think of a better word.

Author: I changed it slightly and used a better word.
